internal/datastore/postgres/snapshot.go

```diff
@@ -0,0 +1,236 @@
+// Adapted from https://github.com/jackc/pgtype/blob/f59f1408937ef0bed249f2bfbafb77222bb48f65/xid.go
+
+package postgres
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/jackc/pgtype"
+	"golang.org/x/exp/slices"
+)
+
+type pgSnapshot struct {
+	xmin, xmax uint64
+	xipList    []uint64 // Must always be sorted
+
+	status pgtype.Status
+}
+
+func (s *pgSnapshot) DecodeText(ci *pgtype.ConnInfo, src []byte) error {
+	s.status = pgtype.Undefined
+	asText := string(src)
+	components := strings.SplitN(asText, ":", 3)
+	if len(components) != 3 {
+		return fmt.Errorf("wrong number of snapshot components: %s", asText)
+	}
+
+	var err error
+	s.xmin, err = strconv.ParseUint(components[0], 10, 64)
+	if err != nil {
+		return fmt.Errorf("unable to parse xmin: %s", components[0])
+	}
+
+	s.xmax, err = strconv.ParseUint(components[1], 10, 64)
+	if err != nil {
+		return fmt.Errorf("unable to parse xmax: %s", components[1])
+	}
+
+	if components[2] != "" {
+		xipStrings := strings.Split(components[2], ",")
+		s.xipList = make([]uint64, len(xipStrings))
+		for i, xipStr := range xipStrings {
+			s.xipList[i], err = strconv.ParseUint(xipStr, 10, 64)
+			if err != nil {
+				return fmt.Errorf("unable to parse xip: %s", xipStr)
+			}
+		}
+		slices.Sort(s.xipList)
```

Comment:
when would this not be sorted?

Reply:
This is defensive, it should always be sorted.